File: mock_server.py
"""Mock OpenAI/Anthropic server for testing cache protocol."""
import json
import time
from flask import Flask, request, Response
from flask_cors import CORS

app = Flask(__name__)
CORS(app)

def stream_response(content: str, delay_ms: int = 50):
    """Stream a response chunk by chunk."""
    words = content.split()
    for i, word in enumerate(words):
        chunk = {
            "choices": [{
                "delta": {"content": word + " "},
                "index": 0,
                "finish_reason": None
            }]
        }
        yield f"data: {json.dumps(chunk)}\n\n"
        time.sleep(delay_ms / 1000.0)
    
    # Final chunk
    final = {
        "choices": [{
            "delta": {},
            "index": 0,
            "finish_reason": "stop"
        }]
    }
    yield f"data: {json.dumps(final)}\n\n"
    yield "data: [DONE]\n\n"

@app.route('/v1/chat/completions', methods=['POST'])
def chat_completions():
    """Mock OpenAI chat completions endpoint."""
    data = request.json
    messages = data.get('messages', [])
    
    # Extract the last user message
    user_msg = "Hello"
    for msg in reversed(messages):
        if msg.get('role') == 'user':
            user_msg = msg.get('content', '')[:50]
            break
    
    # Generate a response based on the message
    response_text = f"Mock response to: {user_msg}. This is a streaming test with multiple words to simulate real generation."
    
    print(f"[mock] Received request with {len(messages)} messages, responding with stream")
    
    return Response(
        stream_response(response_text, delay_ms=30),
        mimetype='text/event-stream',
        headers={
            'Cache-Control': 'no-cache',
            'X-Accel-Buffering': 'no'
        }
    )

@app.route('/v1/messages', methods=['POST'])
def anthropic_messages():
    """Mock Anthropic messages endpoint."""
    data = request.json
    messages = data.get('messages', [])
    
    # Extract the last user message
    user_msg = "Hello"
    for msg in reversed(messages):
        if msg.get('role') == 'user':
            content = msg.get('content', '')
            if isinstance(content, list):
                for block in content:
                    if block.get('type') == 'text':
                        user_msg = block.get('text', '')[:50]
                        break
            else:
                user_msg = content[:50]
            break
    
    response_text = f"Mock Anthropic response to: {user_msg}. This simulates Claude streaming with proper event formatting."
    
    print(f"[mock] Received Anthropic request with {len(messages)} messages, responding with stream")
    
    def anthropic_stream():
        words = response_text.split()
        for word in words:
            event = {
                "type": "content_block_delta",
                "delta": {"type": "text_delta", "text": word + " "}
            }
            yield f"data: {json.dumps(event)}\n\n"
            time.sleep(0.03)
        
        yield f"data: {json.dumps({'type': 'message_delta', 'delta': {'stop_reason': 'end_turn'}})}\n\n"
    
    return Response(
        anthropic_stream(),
        mimetype='text/event-stream',
        headers={
            'Cache-Control': 'no-cache',
            'X-Accel-Buffering': 'no'
        }
    )

if __name__ == '__main__':
    print("Starting mock LLM server on http://localhost:5001")
    print("OpenAI endpoint: POST http://localhost:5001/v1/chat/completions")
    print("Anthropic endpoint: POST http://localhost:5001/v1/messages")
    app.run(port=5001, debug=True, threaded=True)

File: mock-server.js
const express = require('express');
const app = express();
app.use(express.json());

// Track requests to simulate different delays
let requestCount = 0;

app.post('/chat/completions', async (req, res) => {
    requestCount++;
    const requestId = requestCount;
    
    console.log(`[${new Date().toISOString()}] Request ${requestId} received`);
    
    const { messages, model, stream } = req.body;
    
    if (!messages || !Array.isArray(messages)) {
        return res.status(400).json({ error: 'messages array is required' });
    }
    
    // Determine delay based on the content
    let delay = 3000; // Default 3s for step 1
    const lastMessage = messages[messages.length - 1]?.content || '';
    
    if (lastMessage.includes('4 to 6')) {
        delay = 6000; // 6s for step 2
        console.log(`[${new Date().toISOString()}] Request ${requestId}: Step 2 detected, using 6s delay`);
    } else if (lastMessage.includes('7 to 9')) {
        delay = 9000; // 9s for step 3
        console.log(`[${new Date().toISOString()}] Request ${requestId}: Step 3 detected, using 9s delay`);
    } else if (lastMessage.includes('Combine')) {
        delay = 1000; // 1s for step 4 (fast)
        console.log(`[${new Date().toISOString()}] Request ${requestId}: Step 4 detected, using 1s delay`);
    } else {
        console.log(`[${new Date().toISOString()}] Request ${requestId}: Step 1 detected, using 3s delay`);
    }
    
    if (stream) {
        res.setHeader('Content-Type', 'text/event-stream');
        res.setHeader('Cache-Control', 'no-cache');
        res.setHeader('Connection', 'keep-alive');
        
        const chunks = [
            { content: 'Chunk 1 ' },
            { content: 'of response ' },
            { content: `after ${delay/1000}s delay. ` },
            { content: 'This is chunk 2. ' },
            { content: 'Final chunk.' }
        ];
        
        console.log(`[${new Date().toISOString()}] Request ${requestId}: Starting stream with ${delay}ms total delay`);
        const startTime = Date.now();
        
        // Send chunks with timing
        for (let i = 0; i < chunks.length; i++) {
            const chunkDelay = (delay / chunks.length) * (i + 1);
            const elapsed = Date.now() - startTime;
            const waitTime = chunkDelay - elapsed;
            
            if (waitTime > 0) {
                await new Promise(resolve => setTimeout(resolve, waitTime));
            }
            
            const chunk = {
                id: `chatcmpl-${requestId}-${i}`,
                object: 'chat.completion.chunk',
                created: Math.floor(Date.now() / 1000),
                model: model || 'gpt-4o-mini',
                choices: [{
                    index: 0,
                    delta: i === chunks.length - 1 ? {} : { content: chunks[i].content },
                    finish_reason: i === chunks.length - 1 ? 'stop' : null
                }]
            };
            
            res.write(`data: ${JSON.stringify(chunk)}\n\n`);
            console.log(`[${new Date().toISOString()}] Request ${requestId}: Sent chunk ${i + 1}/${chunks.length}`);
        }
        
        res.write('data: [DONE]\n\n');
        res.end();
        
        const totalTime = Date.now() - startTime;
        console.log(`[${new Date().toISOString()}] Request ${requestId}: Stream complete in ${totalTime}ms`);
    } else {
        // Non-streaming response
        await new Promise(resolve => setTimeout(resolve, delay));
        
        const response = {
            id: `chatcmpl-${requestId}`,
            object: 'chat.completion',
            created: Math.floor(Date.now() / 1000),
            model: model || 'gpt-4o-mini',
            choices: [{
                index: 0,
                message: {
                    role: 'assistant',
                    content: `Response after ${delay/1000}s delay.`
                },
                finish_reason: 'stop'
            }],
            usage: {
                prompt_tokens: 10,
                completion_tokens: 10,
                total_tokens: 20
            }
        };
        
        res.json(response);
        console.log(`[${new Date().toISOString()}] Request ${requestId}: Non-stream response sent after ${delay}ms`);
    }
});

const PORT = 5001;
app.listen(PORT, () => {
    console.log(`Mock OpenAI server running on http://127.0.0.1:${PORT}`);
    console.log('Timing rules:');
    console.log('  - Step 1 (default): 3s');
    console.log('  - Step 2 (4 to 6): 6s');
    console.log('  - Step 3 (7 to 9): 9s');
    console.log('  - Step 4 (Combine): 1s');
});
